### PR TITLE
BLS12-377 Groth16 verifier contract code

### PR DESCRIPTION
**Pull request by dtebbs**
- contract library
- mixer and client refactors to try and encapsulate curve-specific code

### CODE REVIEW

File: zeth_contracts/contracts/bls12_377_groth16.sol [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+            mstore(add(pad, 0x380), mload(add(proof, 0x080)))
+            mstore(add(pad, 0x3a0), mload(add(proof, 0x0a0)))
+            mstore(add(pad, 0x3c0), mload(add(proof, 0x0c0)))
+            mstore(add(pad, 0x3e0), mload(add(proof, 0x0e0)))
+            mstore(add(pad, 0x400), mload(add(proof, 0x100)))
+            mstore(add(pad, 0x420), mload(add(proof, 0x120)))
+            mstore(add(pad, 0x440), mload(add(proof, 0x140)))
+            mstore(add(pad, 0x460), mload(add(proof, 0x160)))
+
+            // write proof.c, followed by vk.minus_delta to offset 0x480~
```
Comment (AntoineRondelet) [inline zeth_contracts/contracts/bls12_377_groth16.sol:240]: What is the `~` at the end of the line? Do I understand correctly that this means "onward"? Same question for the lines above

File: zeth_contracts/contracts/bls12_377_groth16.sol [hunk 2]
```diff
@@ -0,0 +1,261 @@
+// Copyright (c) 2015-2020 Clearmatics Technologies Ltd
+//
+// SPDX-License-Identifier: LGPL-3.0+
+
+pragma solidity ^0.5.0;
+
+library bls12_377_groth16
```
Comment (AntoineRondelet) [inline zeth_contracts/contracts/bls12_377_groth16.sol:7]: naming style inconsistent, CamelCase is used for other contracts and library code

File: zeth_contracts/contracts/bls12_377_groth16.sol (hunk 1 above)
Comment (dtebbs) [reply zeth_contracts/contracts/bls12_377_groth16.sol:240]: Yes.  I've used it in `0x300~0x380` to mean an interval, and in `0x300~` to mean onward.  I assumed it was pretty clear and intuitive, but if not we can change it.

File: zeth_contracts/contracts/Groth16Mixer.sol [hunk 3]
```diff
@@ -7,92 +7,38 @@ pragma experimental ABIEncoderV2;
 
 import "./OTSchnorrVerifier.sol";
 import "./BaseMixer.sol";
-import "./Pairing.sol";
+import "./alt_bn128_groth16.sol";
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply zeth_contracts/contracts/Groth16Mixer.sol]: renamed to `Groth1AltBN128`

File: zeth_contracts/contracts/bls12_377_groth16.sol (hunk 2 above)
Comment (dtebbs) [reply zeth_contracts/contracts/bls12_377_groth16.sol]: renamed to `Groth16BLS12_377`

File: client/zeth/core/signing.py [hunk 4]
```diff
@@ -15,9 +15,16 @@
...
-from typing import List
+from typing import List, Tuple
+
+FQ = ec.FQ
+G1 = Tuple[ec.FQ, ec.FQ]
+
+# pylint: disable=line-too-long
+# Order of the subgroup used here.
+SIGNATURE_PRIME = \
+    21888242871839275222246405745257275088548364400416034343698204186575808495617  # noqa
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply client/zeth/core/signing.py:26]: Expanded the comment, referring to the fuller description at the top of the file

[the remark this answers is not in the document]
Comment (AntoineRondelet) [reply client/zeth/core/signing.py:26]: Thanks

File: zeth_contracts/contracts/bls12_377_groth16.sol (hunk 1 above)
Comment (AntoineRondelet) [reply zeth_contracts/contracts/bls12_377_groth16.sol]: Thanks, was keen to double check to make sure I got it. Looks clear enough to me.

**Review by AntoineRondelet — APPROVED** [review]
LGTM

**Pull request merged**